### PR TITLE
Update SSD example to use distributed JoC model

### PR DESCRIPTION
**Pull request by awolant**
Signed-off-by: Albert Wolant <awolant@nvidia.com>

### CODE REVIEW

File: docs/examples/pytorch/single_stage_detector/LICENSE [hunk 1]
```diff
@@ -185,6 +185,8 @@
       file or class name and description of purpose be included on the
       same "printed page" as the copyright notice for easier
       identification within third-party archives.
+   
+   Copyright 2018 NVIDIA Corporation
```
Comment (Kh4L) [inline docs/examples/pytorch/single_stage_detector/LICENSE]: The whole code base was written solely in 2018?

File: docs/examples/pytorch/single_stage_detector/pytorch_ssd.rst [hunk 2]
```diff
@@ -4,7 +4,7 @@ Single Shot MultiBox Detector training in PyTorch
 This example shows how DALI can be used in detection networks, specifically Single Shot Multibox Detector originally published by Wei Liu, Dragomir Anguelov, Dumitru Erhan, Christian Szegedy, 
 Scott Reed, Cheng-Yang Fu, Alexander C. Berg as `SSD: Single Shot MultiBox Detector <https://arxiv.org/abs/1512.02325>`_.
 
-Code is based on `MLPerf example <https://github.com/mlperf/training/tree/master/single_stage_detector/ssd>`_ and has been modified to use DALI. 
+Code is based on `NVIDIA Deep Learining Examples <https://github.com/awolant/DeepLearningExamples/tree/master/PyTorch/Detection/SSD>`_ and has been modified to use full DALI pipeline.
```
Comment (Kh4L) [inline docs/examples/pytorch/single_stage_detector/pytorch_ssd.rst]:
```suggestion
Code is based on `NVIDIA Deep Learning Examples <https://github.com/awolant/DeepLearningExamples/tree/master/PyTorch/Detection/SSD>`_ and has been modified to use full DALI pipeline. 
```

Comment (awolant) [reply docs/examples/pytorch/single_stage_detector/pytorch_ssd.rst]: Done

File: docs/examples/pytorch/single_stage_detector/LICENSE (hunk 1 above)
Comment (awolant) [reply docs/examples/pytorch/single_stage_detector/LICENSE]: Done

**Comment by awolant** [comment]
Build 700635 passed with L1

File: docs/examples/pytorch/single_stage_detector/main.py [hunk 3]
```diff
@@ -1,55 +1,215 @@
...
+    iteration = 0
+    loss_func = Loss(dboxes)
+
+    loss_func.cuda()
+
+    optimizer = torch.optim.SGD(
+        tencent_trick(ssd300), 
+        lr=args.learning_rate,
+        momentum=args.momentum,
+         weight_decay=args.weight_decay)
```
Comment (jantonguirao) [inline docs/examples/pytorch/single_stage_detector/main.py]: indent

File: docs/examples/pytorch/single_stage_detector/pytorch_ssd.rst [hunk 4]
```diff
@@ -4,7 +4,7 @@ Single Shot MultiBox Detector training in PyTorch
 This example shows how DALI can be used in detection networks, specifically Single Shot Multibox Detector originally published by Wei Liu, Dragomir Anguelov, Dumitru Erhan, Christian Szegedy, 
 Scott Reed, Cheng-Yang Fu, Alexander C. Berg as `SSD: Single Shot MultiBox Detector <https://arxiv.org/abs/1512.02325>`_.
 
-Code is based on `MLPerf example <https://github.com/mlperf/training/tree/master/single_stage_detector/ssd>`_ and has been modified to use DALI. 
+Code is based on `NVIDIA Deep Learning Examples <https://github.com/awolant/DeepLearningExamples/tree/master/PyTorch/Detection/SSD>`_ and has been modified to use full DALI pipeline.
```
Comment (jantonguirao) [inline docs/examples/pytorch/single_stage_detector/pytorch_ssd.rst]: intentional?

File: docs/examples/pytorch/single_stage_detector/src/model.py [hunk 5]
```diff
@@ -263,12 +189,12 @@ def forward(self, ploc, plabel, gloc, glabel):
         neg_num = torch.clamp(3*pos_num, max=mask.size(1)).unsqueeze(-1)
         neg_mask = con_rank < neg_num
 
+        #print(con.shape, mask.shape, neg_mask.shape)
```
Comment (jantonguirao) [inline docs/examples/pytorch/single_stage_detector/src/model.py]: should stay?

File: docs/examples/pytorch/single_stage_detector/pytorch_ssd.rst (hunk 4 above)
Comment (awolant) [reply docs/examples/pytorch/single_stage_detector/pytorch_ssd.rst]: Fixed

File: docs/examples/pytorch/single_stage_detector/main.py [hunk 6]
```diff
@@ -1,55 +1,215 @@
...
+
+    scheduler = MultiStepLR(
+        optimizer=optimizer, 
+        milestones=args.multistep, 
+        gamma=0.1)
+
+    if args.fp16:
+        if args.amp:
```
Comment (mzient) [inline docs/examples/pytorch/single_stage_detector/main.py]: Maybe `amp` should imply `fp16`?

File: docs/examples/pytorch/single_stage_detector/main.py [hunk 7]
```diff
@@ -1,55 +1,215 @@
...
+    parser.add_argument(
+        '--weight-decay', '--wd', type=float, default=0.0005,
+        help='momentum argument for SGD optimizer')
     parser.add_argument('--warmup', type=int, default=None)
+    parser.add_argument(
+        '--backbone', type=str, default='resnet50',
+        choices=['resnet18', 'resnet34', 'resnet50', 'resnet101', 'resnet152'])
     parser.add_argument('--num-workers', type=int, default=4)
-    parser.add_argument('--fp16', action='store_true')
+    parser.add_argument('--fp16', action='store_true', help='use halp precission')
```
Comment (mzient) [inline docs/examples/pytorch/single_stage_detector/main.py]: Suggestion --fp16 = off|amp|static?

Comment (awolant) [reply docs/examples/pytorch/single_stage_detector/main.py]: Done

File: docs/examples/pytorch/single_stage_detector/main.py (hunk 6 above)
Comment (awolant) [reply docs/examples/pytorch/single_stage_detector/main.py]: Done

File: docs/examples/pytorch/single_stage_detector/src/model.py (hunk 5 above)
Comment (awolant) [reply docs/examples/pytorch/single_stage_detector/src/model.py]: Done

File: docs/examples/pytorch/single_stage_detector/main.py (hunk 3 above)
Comment (awolant) [reply docs/examples/pytorch/single_stage_detector/main.py]: Done

**Pull request merged**